[JSC] add support for `Iterator.concat`

dcrousso opened the pull request:
#### e2cc521c373c50c6048391237658c08a00973f35
<pre>
[JSC] add support for `Iterator.concat`
<a href="https://bugs.webkit.org/show_bug.cgi?id=281109">https://bugs.webkit.org/show_bug.cgi?id=281109</a>

Reviewed by Alexey Shvayka.

This will allow for easier composition of iterators instead of having to manually create a generator or spread into an array.

Proposal: &lt;<a href="https://github.com/tc39/proposal-iterator-sequencing">https://github.com/tc39/proposal-iterator-sequencing</a>&gt;
Spec: &lt;<a href="https://tc39.es/proposal-iterator-sequencing/">https://tc39.es/proposal-iterator-sequencing/</a>&gt;

* Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
(concat): Added.
* Source/JavaScriptCore/runtime/JSIteratorConstructor.cpp:
(JSC::JSIteratorConstructor::finishCreation):
* Source/JavaScriptCore/runtime/OptionsList.h:

* Source/JavaScriptCore/builtins/JSIteratorHelperPrototype.js:
(return):
Check if the `UnderlyingIterator` is null (in the case of `Iterator.concat`) before attempting to close it.

* JSTests/stress/iterator-concat.js: Added.

Canonical link: <a href="https://commits.webkit.org/285587@main">https://commits.webkit.org/285587@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bffc693be31e682adf6fea8689853e6883514807

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73154 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52583 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25962 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77375 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24392 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60388 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/365 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57490 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15963 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47508 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62955 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37908 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20432 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22721 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66289 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66000 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20786 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79036 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72411 "Built successfully and passed tests") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/53 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65918 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65200 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16116 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7199 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94190 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/433 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20713 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/462 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/447 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/464 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,44 @@ function from(value)
+            @throwTypeError("Iterator.concat expects all arguments to be objects");
+
+        var openMethod = iterable.@@iterator;
+        if (@isUndefinedOrNull(openMethod))
```
According to the spec,

> c. If method is undefined, throw a TypeError exception.

so we should check only `undefined`? (Or feedback to the spec)

dcrousso replied:
i was matching what was done for some of the other iterator helper methods, so if it truly is that we only want to check for `undefined` then we probably need to update those too (or i could be incorrectly reading the spec for those methods)

michaelficarra replied:
GetMethod returns `undefined` for both `null` and `undefined`. See https://tc39.es/ecma262/multipage/abstract-operations.html#sec-getmethod.

It also checks for callability and throws if the value isn't callable.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSIteratorConstructor.cpp:
```diff
@@ -56,6 +56,9 @@ void JSIteratorConstructor::finishCreation(VM& vm, JSGlobalObject* globalObject,
     putDirectWithoutTransition(vm, vm.propertyNames->prototype, iteratorPrototype, static_cast<unsigned>(PropertyAttribute::DontEnum | PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly));
     JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->from, jsIteratorConstructorFromCodeGenerator, static_cast<unsigned>(PropertyAttribute::DontEnum));
+
+    if (Options::useIteratorHelpers())
```
Instead of using useIteratorHelpers, I think we should create a new feature flag option such as useIteratorSequencing.

dcrousso replied:
i can make a separate flag if you think it's needed but i felt it was close enough to the other iterator methods that it'd be find to group together

btw i think this was originally part of iterator helpers but later split out to discuss separately and it also relies on a few methods defined for iterator helpers, so it probably would have to be guarded by that flag anyways

shvaikalesh replied:
+1 on separate flag, we are about to enable `--useIteratorHelpers` by default

sosukesuzuki commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,44 @@ function from(value)
+        }
+    })();
+
+    return @iteratorHelperCreate(generator, null);
```
FYI: According to the spec, if the `IteratorHelperPrototype` object doesn't have an `[[UnderlyingIterator]]`, the return method should throw an error, but I wasn't sure if this was the intended behavior, so I created an issue in the spec: https://github.com/tc39/proposal-iterator-sequencing/issues/14

dcrousso replied:
oh neat yeah i was gonna ask a similar question but then saw that someone (which i now know was you) already did.  thanks! :)

shvaikalesh replied:
IteratorHelper's `return` will throw a TypeError if `[[UnderlyingIterator]]` is `null`. Please add a `null` check there and a test.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,44 @@ function from(value)
+    }
+
+    var generator = (function*() {
+        for (var iterable of iterables) {
```
This is observable (reliant on `Array.prototype[Symbol.iterator]`), please use old-school `for` loop.

Also I would consider using two arrays rather than array of objects.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,47 @@ function from(value)
+    var openMethods = [];
+    var iterables = [];
+    for (var i = 0; i < count; ++i) {
+        var iterable = arguments[i];
```
```suggestion
        var iterable = @argument(i);
```

dcrousso replied:
i dont think this works.  every other use of `@arguments()` passes in a constant, and there's lots of places that use `arguments[]`.  when i tried this it kept giving me the first argument over and over, regardless of the value of `i`

shvaikalesh replied:
Oh, my bad, `@argument` requires a constant.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,47 @@ function from(value)
+        var openMethod = iterable.@@iterator;
+        if (@isUndefinedOrNull(openMethod))
+            @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
```
Please add `@isCallable` check here: even tho `@call` with throw just the same TypeError in case of non-callable value, we should avoid performing `[[Get]]` on following arguments, as it is observable. It would be great to have a test for that.

shvaikalesh replied:
> It would be great to have a test for that.

in the test262 suite sometime later

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,47 @@ function from(value)
+{
+    "use strict";
+
+    var count = @argumentCount();
```
```suggestion
    var argumentCount = @argumentCount();
```

to avoid similarities with counters we pass to callback in other methods?

shvaikalesh reviewed (changes requested):
We are getting there!

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,47 @@ function from(value)
+                @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
+            for (;;) {
+                var result = @iteratorGenericNext(iterator.next, iterator);
```
Please hoist `iterator.next` out of the loop, it is observable and per spec performed only once per iterator.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -71,3 +71,49 @@ function from(value)
+            @throwTypeError("Iterator.concat expects all arguments to be objects");
+
+        var openMethod = iterable.@@iterator;
+        if (@isUndefinedOrNull(openMethod) || !@isCallable(openMethod))
```
We can simplify to:
```suggestion
        if (!@isCallable(openMethod))
```

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorHelperPrototype.js:
```diff
@@ -51,7 +51,12 @@ function return()
+
+        var underlyingIterators = @getIteratorHelperInternalField(this, @iteratorHelperFieldUnderlyingIterators);
+        if (underlyingIterators) {
+            for (var i = 0; i < underlyingIterators.length; ++i)
```
Please allow me a business day or two to think about this!

shvaikalesh replied:
@dcrousso Please note that https://tc39.es/proposal-joint-iteration/#sec-closeall, apart from going in reverse order, is carefully delaying abrupt completions of `return` lookups / calls from throwing an exception until all iterators are closed. It's not easy to implement in JS.

Let's take simpler approach here by just null-checking the `@getIteratorHelperInternalField(this, @iteratorHelperFieldUnderlyingIterators)`.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorHelperPrototype.js:
```diff
@@ -51,7 +51,10 @@ function return()
-        @iteratorGenericClose(@getIteratorHelperInternalField(this, @iteratorHelperFieldUnderlyingIterator));
+
+        var underlyingIterator = @getIteratorHelperInternalField(this, @iteratorHelperFieldUnderlyingIterator);
+        if (!@isUndefinedOrNull(underlyingIterator))
```
Let's make the check a bit more precise:

```suggestion
        if (underlyingIterator !== null)
```

shvaikalesh reviewed (approved):
r=me with nit

dcrousso committed:
[JSC] add support for `Iterator.concat`
https://bugs.webkit.org/show_bug.cgi?id=281109

Reviewed by Alexey Shvayka.

This will allow for easier composition of iterators instead of having to manually create a generator or spread into an array.

Proposal: <https://github.com/tc39/proposal-iterator-sequencing>
Spec: <https://tc39.es/proposal-iterator-sequencing/>

* Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
(concat): Added.
* Source/JavaScriptCore/runtime/JSIteratorConstructor.cpp:
(JSC::JSIteratorConstructor::finishCreation):
* Source/JavaScriptCore/runtime/OptionsList.h:

* Source/JavaScriptCore/builtins/JSIteratorHelperPrototype.js:
(return):
Check if the `UnderlyingIterator` is null (in the case of `Iterator.concat`) before attempting to close it.

* JSTests/stress/iterator-concat.js: Added.

Canonical link: https://commits.webkit.org/285587@main
